### PR TITLE
[WIP] Preventing multiple deployments from running concurrently

### PR DESCRIPTION
**Pull request by abhgupta**
Additionally, this PR also moves some of the duplicated code into the util file.

### CODE REVIEW

**Comment by abhgupta** [comment]
[test]

**Comment by abhgupta** [comment]
In progress:
 - Adding a wait/sleep before re-queuing a deployment when a previous deployment is still in new/pending/running state.

File: test/integration/deploy_trigger_test.go [hunk 1]
```diff
@@ -230,6 +222,27 @@ waitForPendingStatus:
...
+	deployment.Annotations[deployapi.DeploymentStatusAnnotation] = string(deployapi.DeploymentStatusComplete)
+	// update the deployment
+	if _, err = kubeClusterAdminClient.ReplicationControllers(testutil.Namespace()).Update(deployment); err != nil {
+		t.Fatalf("Error updating existing deployment: %v", err)
+	}
+
+	event := <-watch.ResultChan()
+	if e, a := watchapi.Modified, event.Type; e != a {
+		t.Fatalf("expected watch event type %s, got %s", e, a)
+	}
```
Comment (abhgupta) [inline test/integration/deploy_trigger_test.go]: This integration test seems a bit fragile with the way its structured currently to use a watch. I will (either now or later as a separate PR) modify it to not use a watch for additions/modifications to replication controllers.

File: pkg/deploy/controller/deploymentconfig/controller.go [hunk 2]
```diff
@@ -59,10 +65,30 @@ func (c *DeploymentConfigController) Handle(config *deployapi.DeploymentConfig)
 		}
 	}
 
+	// Check if any previous deployment is still running (any non-terminal state).
+	existingDeployments, err := c.deploymentClient.listDeploymentsForConfig(config.Namespace, config.Name)
+	if err != nil {
+		return fmt.Errorf("couldn't list deployments for config %s: %v", deployutil.LabelForDeploymentConfig(config), err)
+	} else {
+		for _, deployment := range existingDeployments.Items {
```
Comment (ironcladlou) [inline pkg/deploy/controller/deploymentconfig/controller.go]: The `if` returns; no need for the `else`

File: pkg/deploy/controller/deploymentconfig/controller.go [hunk 3]
```diff
@@ -59,10 +65,30 @@ func (c *DeploymentConfigController) Handle(config *deployapi.DeploymentConfig)
 		}
 	}
 
+	// Check if any previous deployment is still running (any non-terminal state).
+	existingDeployments, err := c.deploymentClient.listDeploymentsForConfig(config.Namespace, config.Name)
+	if err != nil {
+		return fmt.Errorf("couldn't list deployments for config %s: %v", deployutil.LabelForDeploymentConfig(config), err)
+	} else {
```
Comment (ironcladlou) [inline pkg/deploy/controller/deploymentconfig/controller.go]: The `if` returns; no need for the `else`

File: pkg/deploy/controller/deploymentconfig/controller.go [hunk 4]
```diff
@@ -59,10 +65,30 @@ func (c *DeploymentConfigController) Handle(config *deployapi.DeploymentConfig)
...
+		for _, deployment := range existingDeployments.Items {
+			deploymentStatus := deployutil.StatusForDeployment(&deployment)
+			switch deploymentStatus {
+			case deployapi.DeploymentStatusFailed,
+				deployapi.DeploymentStatusComplete:
+				// Previous deployment in terminal state - can ignore
+				// Ignoring specific deployment states so that any new
+				// deployment state will not be ignored
+			default:
+				glog.V(4).Infof("Found previous deployment %s/%s (status %s) - will requeue", deployment.Namespace, deployment.Name, deploymentStatus)
```
Comment (ironcladlou) [inline pkg/deploy/controller/deploymentconfig/controller.go]: Use your new `LabelForDeployment`

File: pkg/deploy/controller/deploymentconfig/controller_test.go [hunk 5]
```diff
@@ -245,3 +246,148 @@ func TestHandle_fatalError(t *testing.T) {
...
+			listDeploymentsForConfigFunc: func(namespace, configName string) (*kapi.ReplicationControllerList, error) {
+				return existingDeployments, nil
+			},
+		},
+		recorder: &record.FakeRecorder{},
+	}
+
+	type existing struct {
+		version int
+		status  string
```
Comment (ironcladlou) [inline pkg/deploy/controller/deploymentconfig/controller_test.go]: You already convert these with `string()` in the scenario loop; make this type `DeploymentStatus` and then you can remove all the `string()` conversion in your scenario list (would look a little cleaner).

File: pkg/deploy/controller/deploymentconfig/controller_test.go [hunk 6]
```diff
@@ -245,3 +246,148 @@ func TestHandle_fatalError(t *testing.T) {
...
+
+		if err != nil {
+			t.Fatalf("unexpected error: %v", err)
+		}
+	}
+}
+
+// TestHandle_previousCompletedDeployment ensures that an attempt to create a
+// new deployment for a config for which all the previous deployments are
+// in a terminal state (complete, failed) is successful
```
Comment (ironcladlou) [inline pkg/deploy/controller/deploymentconfig/controller_test.go]: Comment is inaccurate

File: pkg/deploy/controller/deploymentconfig/controller_test.go [hunk 7]
```diff
@@ -245,3 +246,148 @@ func TestHandle_fatalError(t *testing.T) {
 		t.Fatalf("expected a fatal error, got: %v", err)
 	}
 }
+
+// TestHandle_previousCompletedDeployment ensures that an attempt to create a
+// new deployment for a config for which all the previous deployments are
+// in a terminal state (complete, failed) is successful
+func TestHandle_previousCompletedDeployment(t *testing.T) {
```
Comment (ironcladlou) [inline pkg/deploy/controller/deploymentconfig/controller_test.go]: I think you can combine this test with `TestHandle_previousPendingDeployment` by adding an `error` field to the `scenario`; then the scenarios which have terminal deployments would have a `nil` error expectation, and the non-terminal scenarios could have a `transientError` expectation.

**Comment by abhgupta** [comment]
@ironcladlou Implemented feedback - PTAL

**Comment by abhgupta** [comment]
@ironcladlou After this PR is merged, I have the changes that detect multiple inflight deployments and will hook into the stop/cancel API once its available.

**Comment by ironcladlou** [comment]
My only remaining issue here is that we're going to be effectively tight-loop processing transient errors given the current retry mechanics. I think we should probably come up with *something*, even if it's not ideal in the long term.

**Comment by abhgupta** [comment]
@ironcladlou The current rate limiter serves exactly this purpose - it ensures that the retries don't happen in a tight loop. As of now, retries will happen (resources will be re-queued on error) 1s apart (excluding bursts). This is configurable and we can modify/tweak the rate limiter parameters. However, this applies across the board for ALL errors and does not distinguish between the types of errors.

**Comment by ironcladlou** [comment]
>The current rate limiter serves exactly this purpose - it ensures that the retries don't happen in a tight loop. As of now, retries will happen (resources will be re-queued on error) 1s apart (excluding bursts). This is configurable and we can modify/tweak the rate limiter parameters. However, this applies across the board for ALL errors and does not distinguish between the types of errors.

Ah yes. And it applies to the entire queue, correct (e.g. repeated attempts to retry the deployment will slow down processing of other deployments)? Let's live with it for now and optimize later.

This LGTM; I'll check in with you today about merging.

**Comment by abhgupta** [comment]
The rate limiting only applies to retries being requeued. Normal processing of deployments is not rate limited. If an error occurs, a token is required to be available before requeuing the deployment. The token check is a blocking call and returns only when a token is available. Based on current parameters, a new token is added every 1s and upto 10 tokens can be accumulated for a burst.

**Comment by ironcladlou** [comment]
We're satisfied with rate limiting as-is for now, [merge]

**Pull request merged**